code/game/objects/effects/effect_system/effects_foam.dm

```diff
@@ -2,164 +2,254 @@
+	if(!target_turf || !target_turf.Enter(src))
+		return
+
+	var/obj/effect/particle_effect/foam/new_foam
 
-/obj/effect/particle_effect/foam/process()
-	if(--amount < 0)
+	new_foam = locate() in target_turf
+	if(new_foam)
 		return
 
+	new_foam = new type(target_turf)
+	new_foam.spread_amount = spread_amount
+	new_foam.spread_time = spread_time
+	new_foam.solidify_time = solidify_time
+	new_foam.max_amount_on_spread = max_amount_on_spread
+	new_foam.spread_at_range = spread_at_range
+	new_foam.react_mode = react_mode
+	new_foam.max_reagent_filling = max_reagent_filling
+
+
+	// add the new amount of foam
+	if(reagents)
+		for(var/datum/reagent/R in reagents.reagent_list)
+			new_foam.reagents.add_reagent(R.id, min(R.volume, 5), R.data, reagents.chem_temp)
+		new_foam.color = mix_color_from_reagents(reagents.reagent_list)
+	if(react_mode & FOAM_REACT_BEFORE_SPREAD)
+		new_foam.disperse_reagents()
+
+/obj/effect/particle_effect/foam/proc/spread()
 	for(var/direction in GLOB.cardinal)
+		var/turf/T = get_step(src, direction)
+		try_spread_to(T)
 
-		var/turf/T = get_step(src,direction)
-		if(!T)
-			continue
+/obj/effect/particle_effect/foam/proc/generate_color()
+	color = mix_color_from_reagents(reagents.reagent_list)
 
-		if(!T.Enter(src))
-			continue
+/obj/effect/particle_effect/foam/process()
 
-		var/obj/effect/particle_effect/foam/F = locate() in T
-		if(F)
-			continue
+	if(react_mode & FOAM_REACT_DURING_SPREAD)
+		disperse_reagents()
+
+	if(--spread_amount < 0)
+		return
+
+
+	spread()
 
-		F = new /obj/effect/particle_effect/foam(T, metal)
-		F.amount = amount
-		if(!metal)
-			F.create_reagents(25)
-			if(reagents)
-				for(var/datum/reagent/R in reagents.reagent_list)
-					F.reagents.add_reagent(R.id, min(R.volume, 5), R.data, reagents.chem_temp)
-				F.color = mix_color_from_reagents(reagents.reagent_list)
 
 // foam disolves when heated
 // except metal foams
 /obj/effect/particle_effect/foam/fire_act(datum/gas_mixture/air, exposed_temperature, exposed_volume, global_overlay = TRUE) //Don't heat the reagents inside
 	return
 
 /obj/effect/particle_effect/foam/temperature_expose(datum/gas_mixture/air, exposed_temperature, exposed_volume) // overriden to prevent weird behaviors with heating reagents inside
-	if(!metal && prob(max(0, exposed_temperature - 475)))
+	if(prob(max(0, exposed_temperature - 475)))
 		flick("[icon_state]-disolve", src)
-
-		spawn(5)
-			qdel(src)
+		QDEL_IN(src, 0.5 SECONDS)
 
 /obj/effect/particle_effect/foam/Crossed(atom/movable/AM, oldloc)
-	if(metal)
+	if(!iscarbon(AM))
 		return
+	var/mob/living/carbon/M = AM
+	if((M.slip("foam", 10 SECONDS) || IS_HORIZONTAL(M)) && reagents)
+		fill_with_reagents(M)
+
+
+/obj/effect/particle_effect/foam/metal
+	name = "metal foam"
+	icon_state = "mfoam"  // finally mentor foam
+	spread_time = 1.2 SECONDS
+	react_mode = FOAM_REACT_ON_DISSIPATE
+	/// Represents the icon state that we'll become when we solidify
+	var/metal = METAL_FOAM_ALUMINUM
```

Comment:
I would rename this to "metal_state" or something as it's a bit more descriptive